daemons/ipa-sam/ipa_sam.c

```diff
@@ -2624,13 +2688,20 @@ static NTSTATUS ipasam_set_trusted_domain(struct pdb_methods *methods,
 
 	if (entry == NULL) { /* FIXME: allow password updates here */
 		status = get_trust_pwd(tmp_ctx, &td->trust_auth_incoming,
-				       &trustpw, NULL);
+				       &trustpw_incoming, NULL);
 		if (!NT_STATUS_IS_OK(status)) {
 			goto done;
 		}
-		res = set_cross_realm_princs(ipasam_state, td->domain_name,
-					     trustpw, td->trust_direction);
-		memset(trustpw, 0, strlen(trustpw));
+		status = get_trust_pwd(tmp_ctx, &td->trust_auth_outgoing,
+				       &trustpw_outgoing, NULL);
+		if (!NT_STATUS_IS_OK(status)) {
+			goto done;
+		}
+		res = set_cross_realm_princs(ipasam_state, td->domain_name, td->netbios_name,
+					     trustpw_incoming, trustpw_outgoing,
+					     td->trust_direction);
+		memset(trustpw_incoming, 0, strlen(trustpw_incoming));
```

Comment:
The compiler may optimize the ``memset()`` call away. You need to use C11 ``memset_s`` or some hack with volatile pointers.